sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SetFns.java

```diff
@@ -0,0 +1,618 @@
+   *
+   * <pre>{@code
+   * Pipeline p = ...;
+   * PCollection<String> first = p.apply(Create.of("1","1","1", "2", "3", "3","4", "5"));
+   * PCollection<String> second = p.apply(Create.of("1","1", "3", "4","4", "6"));
+   *
+   * PCollection<String> third = p.apply(Create.of("1", "2","2"));
+   *
+   * // Following example will perform (first intersect second) intersect third.
+   * PCollection<String> results =
+   *     PCollectionList.of(first).and(second).and(third)
+   *     .apply(SetFns.exceptDistinct()); // results will be PCollection<String> containing: "5"
+   *
+   * }</pre>
+   *
+   * @param <T> the type of the elements in the input {@code PCollectionList<T>} and output {@code
+   *     PCollection<T>}s.
+   */
+  public static <T> SetImplCollections<T> exceptDistinct() {
+    SerializableBiFunction<Long, Long, Long> exceptFn =
+        (numberOfElementsinLeft, numberOfElementsinRight) ->
+            numberOfElementsinLeft > 0 && numberOfElementsinRight == 0 ? 1L : 0L;
+    return new SetImplCollections<>(exceptFn);
+  }
+
+  /**
+   * Returns a new {@code PTransform} transform that follows SET ALL semantics to compute the
+   * difference all (exceptAll) with provided {@code PCollection<T>}.
+   *
+   * <p>The argument should not be modified after this is called.
+   *
+   * <p>The elements of the output {@link PCollection} which will follow EXCEPT_ALL Semantics as
+   * follows: Given there are m elements on pipeline which is constructed {@link PCollection} (left)
+   * and n elements on in provided {@link PCollection} (right): - it will output m elements of left
+   * for all elements which are present in left but not in right. - it will output MAX(m - n, 0)
+   * elements of left for all elements which are present in both left and right.
+   *
+   * <p>Note that this transform requires that the {@code Coder} of the all {@code PCollection<T>}
+   * to be deterministic (see {@link Coder#verifyDeterministic()}). If the collection {@code Coder}
+   * is not deterministic, an exception is thrown at pipeline construction time.
+   *
+   * <p>All inputs must have equal {@link WindowFn}s and compatible triggers (see {@link
+   * Trigger#isCompatible(Trigger)}).
+   *
+   * <p>By default, the output {@code PCollection<T>} encodes its elements using the same {@code
+   * Coder} that of {@code PCollection<T>}
+   *
+   * <pre>{@code
+   * Pipeline p = ...;
+   *
+   * PCollection<String> left = p.apply(Create.of("1","1","1", "2", "3", "3","3","4", "5"));
+   * PCollection<String> right = p.apply(Create.of("1", "3", "4","4", "6"));
+   *
+   * PCollection<String> results =
+   *     left.apply(SetFns.exceptAll(right)); // results will be PCollection<String> containing: "1","1","2","3","3","5"
+   * }</pre>
+   *
+   * @param <T> the type of the elements in the input and output {@code PCollection<T>}s.
+   */
+  public static <T> SetImpl<T> exceptAll(PCollection<T> rightCollection) {
```

Comment:
Since SetImpl and SetImplCollection are private classes, users won't be able to assign them to a local variable without converting it to a PTransform which most IDEs won't do for you automatically. Was this your intent?

In other transforms we have made these kinds of classes public but they are usually specific to what they do, so in this case we would have a UnionAll/Intersect/IntersectAll/Distinct/DistinctAll/... PTransform that was public that had whatever internal implementation you want. The alternative would be to return PTransform<PCollection<T>, PCollection<T>> which seems ok.

Reply:
Yes. It is to avoid mutation and create correct abstraction. As this Transform doesn't have any builder like pattern, does taking out a variable makes any sense?

Reply:
Added return value as respective PTransforms.